### PR TITLE
[SPARK-31441] Support duplicated column names for toPandas with arrow execution.

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

This PR is adding support duplicated column names for `toPandas` with Arrow execution.

### Why are the changes needed?

When we execute `toPandas()` with Arrow execution, it fails if the column names have duplicates.

```py
>>> spark.sql("select 1 v, 1 v").toPandas()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/lib/python3.7/site-packages/pyspark/sql/dataframe.py", line 2132, in toPandas
    pdf = table.to_pandas()
  File "pyarrow/array.pxi", line 441, in pyarrow.lib._PandasConvertible.to_pandas
  File "pyarrow/table.pxi", line 1367, in pyarrow.lib.Table._to_pandas
  File "/path/to/lib/python3.7/site-packages/pyarrow/pandas_compat.py", line 653, in table_to_blockmanager
    columns = _deserialize_column_index(table, all_columns, column_indexes)
  File "/path/to/lib/python3.7/site-packages/pyarrow/pandas_compat.py", line 704, in _deserialize_column_index
    columns = _flatten_single_level_multiindex(columns)
  File "/path/to/lib/python3.7/site-packages/pyarrow/pandas_compat.py", line 937, in _flatten_single_level_multiindex
    raise ValueError('Found non-unique column index')
ValueError: Found non-unique column index
```

### Does this PR introduce any user-facing change?

Yes, previously we will face an error above, but after this PR, we will see the result:

```py
>>> spark.sql("select 1 v, 1 v").toPandas()
   v  v
0  1  1
```

### How was this patch tested?

Added and modified related tests.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/pandas/conversion.py [hunk 1]
```diff
@@ -103,13 +103,17 @@ def toPandas(self):
                 try:
                     from pyspark.sql.pandas.types import _check_series_localize_timestamps
                     import pyarrow
-                    batches = self._collect_as_arrow()
+                    # Rename columns to void duplicated column names.
```
Comment (viirya) [inline python/pyspark/sql/pandas/conversion.py:106]: typo: void -> avoid

**Review by viirya — APPROVED** [review]

File: python/pyspark/sql/pandas/conversion.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/sql/pandas/conversion.py:106]: oops, thanks!

**Comment by HyukjinKwon** [comment]
The last comment was comment only. Merged to master and branch-3.0.

**Comment by HyukjinKwon** [comment]
Should we backport this and SPARK-31186, @viirya and @ueshin?

**Comment by viirya** [comment]
@HyukjinKwon Backport to branch-2.4? Ok, sounds good to me. I will prepare a backport of SPARK-31186.

**Comment by ueshin** [comment]
@HyukjinKwon Sure, I'll submit the backport PR after @viirya's is merged.

**Comment by gatorsmile** [comment]
For other reviewers, this PR is the follow-up of https://github.com/apache/spark/pull/28025

**Pull request closed without merging**